packages/client/src/links/wsLink/wsClient/wsConnection/index.ts

```diff
@@ -0,0 +1,141 @@
+import { behaviorSubject } from '@trpc/server/observable';
+import { createWebSocketConnection } from './createWebSocketConnection';
+import type { PingPongOptions } from './pingPong';
+import { setupPingInterval, setupPingListener } from './pingPong';
+
+export interface WebSocketConnectionOptions {
+  WebSocketPonyfill?: typeof WebSocket;
+  promiseUrl: Promise<string>;
+  keepAlive: PingPongOptions & {
+    enabled: boolean;
+  };
+}
+
+/**
+ * Manages a WebSocket connection with support for reconnection, keep-alive mechanisms,
+ * and observable state tracking.
+ */
+export class WsConnection {
+  static connectCount = 0;
+  public id = ++WsConnection.connectCount;
+
+  private readonly WebSocketPonyfill: typeof WebSocket;
+  private readonly promiseUrl: Promise<string>;
+  private readonly keepAliveOpts: WebSocketConnectionOptions['keepAlive'];
+  public readonly wsObservable = behaviorSubject<WebSocket | null>(null);
+
+  constructor(opts: WebSocketConnectionOptions) {
+    this.WebSocketPonyfill = opts.WebSocketPonyfill ?? WebSocket;
+    if (!this.WebSocketPonyfill) {
+      throw new Error(
+        "No WebSocket implementation found - you probably don't want to use this on the server, but if you do you need to pass a `WebSocket`-ponyfill",
+      );
+    }
+
+    this.promiseUrl = opts.promiseUrl;
+    this.keepAliveOpts = opts.keepAlive;
+  }
+
+  public get ws() {
+    return this.wsObservable.get();
+  }
+
+  private set ws(ws) {
+    this.wsObservable.next(ws);
+  }
+
+  /**
+   * Checks if the WebSocket connection is open and ready to communicate.
+   */
+  public isOpen(): this is { ws: WebSocket } {
+    return !!this.ws && this.ws.readyState === this.WebSocketPonyfill.OPEN;
+  }
+
+  /**
+   * Checks if the WebSocket connection is closed or in the process of closing.
+   */
+  public isClosed(): this is { ws: WebSocket } {
+    return (
+      !!this.ws &&
+      (this.ws.readyState === this.WebSocketPonyfill.CLOSING ||
+        this.ws.readyState === this.WebSocketPonyfill.CLOSED)
+    );
+  }
+
+  /**
+   * Manages the WebSocket opening process, ensuring that only one open operation
+   * occurs at a time. Tracks the ongoing operation with `openPromise` to avoid
+   * redundant calls and ensure proper synchronization.
+   *
+   * Sets up the keep-alive mechanism and necessary event listeners for the connection.
+   *
+   * @returns A promise that resolves once the WebSocket connection is successfully opened.
+   */
+  private openPromise: Promise<void> | null = null;
+  public async open() {
+    if (this.openPromise) return this.openPromise;
+
+    this.id = ++WsConnection.connectCount;
+    const { wsPromise, openPromise } = createWebSocketConnection(
+      this.promiseUrl,
+      this.WebSocketPonyfill,
+    );
+    this.openPromise = openPromise;
+    this.ws = await wsPromise;
+
+    setupPingListener(this.ws);
+    if (this.keepAliveOpts.enabled) {
+      setupPingInterval(this.ws, this.keepAliveOpts);
```

Comment:
This I get it's a standalone fn as it's quite long and adds noise, but you can just put it at the top of the same file maybe?

Reply:
Fixed